docs/proposals/0537-introduce-component-type-definitions.md

```diff
@@ -0,0 +1,592 @@
+
+- **Over-abstraction from Kubernetes APIs**: Current component types mask native Kubernetes resources, making the mental model harder for Kubernetes-savvy users to understand OpenChoreo's Components.
+
+- **Limited platform engineer control**: ComponentClasses can only override portions of developer-defined Deployment specs. Platform engineers cannot enforce organizational policies, resource limits, or deployment configurations as templates that developers instantiate.
+
+- **No composition model**: Platform engineers cannot define their own component types by combining Kubernetes primitives, OpenChoreo concepts, and external CRDs of other tools they may be using.
+
+- **No environment-specific overrides**: There is no way to override or patch environment-specific parameters (like resource limits or replica counts) in a structured, validated manner.
+
+As a result, platform engineers lack the control needed to enforce organizational standards, and they may be forced to bypass OpenChoreo entirely if their requirements do not fit into one of OpenChoreo's opinionated component types.
+
+---
+
+## Goals
+
+- **Extensibility**: Platform engineers can model organizational needs without forking or bypassing OpenChoreo. Addons enable cross-cutting capabilities (storage, networking, security) to be reused across component types.
+
+- **Alignment with Kubernetes**: Reduces friction by staying consistent with native APIs, making it easier for Kubernetes-savvy teams to understand and adopt OpenChoreo.
+
+- **Developer Experience**: Developers get a consistent, simplified interface (single Component CRD) with guardrails, while platform engineers retain full control. OpenChoreo's developer abstractions like endpoints, connections and config schemas are retained without restricting PE control.
+
+- **Reusability**: Define ComponentTypeDefinitions and Addons once, compose them in multiple ways across different components.
+
+---
+
+## Non-Goals
+
+- **Replace Kubernetes primitives**: This proposal does not aim to abstract away or replace Kubernetes resources. Instead, it works with them to provide a higher-level interface while maintaining compatibility.
+
+- **Support multiple primary workload resources**: Each ComponentTypeDefinition is limited to a single primary workload resource. The ComponentTypeDefinition must declare a `workloadType` field (one of: `deployment`, `statefulset`, `cronjob`, `job`), and the primary workload resource's `id` must match this `workloadType`. Additional supporting resources (Services, HPAs, PDBs, etc.) can be defined in the resources section and via addons.
+
+---
+
+## Impact
+
+This proposal introduces new CRDs and controller support:
+
+- New CRDs: `ComponentTypeDefinition` (v1alpha1), `Addon` (v1alpha1), and `EnvSettings` (v1alpha1)
+- New `v1alpha2` version of the `Component` CRD with `componentType` and `addons[]` fields
```

Comment:
Shall we use the existing v1alpha2? Since we are in the early stage, we can break the API. Otherwise, it will be difficult to manage these different versions at this stage.

Reply:
Or we increment and immediately deprecate (remove) the old versions to be clear - just in case anyone is actually using the current versions